core/services/gateway/network/wsconnection.go

```diff
@@ -0,0 +1,103 @@
+package network
+
+import (
+	"errors"
+	"sync/atomic"
+
+	"github.com/gorilla/websocket"
+)
+
+// WSConnectionWrapper is a websocket connection abstraction that supports re-connects.
+// I/O is separated from connection management:
+//   - component doing I/O provides read and write channels that never change
+//   - component managing connections can listen to connection-closed events and call Restart()
+//     to swap the underlying connection object
+//
+// The Wrapper can be used by a server expecting long-lived connections from a given client,
+// as well as a client maintaining such long-lived connection with a given server.
+// This fits the Gateway very well as servers accept connections only from a fixed set of nodes
+// and conversely, nodes only connect to a fixed set of servers (Gateways).
+//
+// The concept of "pumps" is borrowed from https://github.com/smartcontractkit/wsrpc
+type WSConnectionWrapper struct {
+	conn atomic.Pointer[websocket.Conn]
+
+	writeCh    <-chan WriteItem
+	readCh     chan<- ReadItem
+	shutdownCh chan struct{}
+}
+
+type ReadItem struct {
+	MsgType int
+	Data    []byte
+}
+
+type WriteItem struct {
+	MsgType int
+	Data    []byte
+	ErrCh   chan error
+}
+
+func NewWSConnectionWrapper(readCh chan<- ReadItem, writeCh <-chan WriteItem) *WSConnectionWrapper {
+	cw := &WSConnectionWrapper{
+		writeCh:    writeCh,
+		readCh:     readCh,
+		shutdownCh: make(chan struct{}),
+	}
+	// write pump runs until Shutdown() is called
+	go cw.writePump()
+	return cw
+}
+
+func (c *WSConnectionWrapper) Shutdown() {
```

Comment:
`Shutdown` does not wait for read/write pump goroutines to quit. As a user I would expect that I can close my read/write/err channels after `Shutdown` is done, but it's not the case. Worst case I will get panic if one of those goroutines will write to a closed channel.
Another issue is that a second call to `Shutdown` will lock forever. Either describe this in the function contract or have a guard. Similar to `conn.Close`, second attempt to call `Shutdown` should be no-op.

Reply:
`sync.WaitGroup` is the best option IMO

Reply:
Added a shutdown guard. Since the channels are now created internally, I don't wait for goroutines. Let me know if you think this is still desirable.